lib/improver/nbhood.py

```diff
@@ -55,107 +59,243 @@ class BasicNeighbourhoodProcessing(object):
-        kernel = np.ones([1 + x * 2 for x in fullranges])
-        n = np.ogrid[tuple([slice(-x, x+1) for x in ranges])]
+        # Define the size of the kernel based on the number of grid cells
+        # contained within the desired radius.
+        kernel = np.ones([int(1 + x * 2) for x in fullranges])
+        # Create an open multi-dimensional meshgrid.
+        meshgrid = np.ogrid[tuple([slice(-x, x+1) for x in ranges])]
         if self.unweighted_mode:
             mask = np.reshape(
-                np.sum([x ** 2 for x in n]) > np.cumprod(ranges)[-1],
+                np.sum([x ** 2 for x in meshgrid]) > np.cumprod(ranges)[-1],
                 np.shape(kernel)
             )
         else:
+            # Create a kernel, such that the central grid point has the
+            # highest weighting, with the weighting decreasing with distance
+            # away from the central grid point.
             kernel[:] = (
-                (np.cumprod(ranges)[-1] - np.sum([x**2. for x in n])) /
+                (np.cumprod(ranges)[-1] - np.sum([x**2. for x in meshgrid])) /
                 np.cumprod(ranges)[-1]
             )
             mask = kernel < 0.
         kernel[mask] = 0.
+        # Smooth the data by applying the kernel.
         cube.data = scipy.ndimage.filters.correlate(
             data, kernel, mode='nearest') / np.sum(kernel)
         return cube
+
+    def process(self, cube):
+        """
+        Calculate a kernel to apply, in order to smooth the input cube.
+
+        Parameters
+        ----------
+        cube : Iris.cube.Cube
+            Cube to apply a kernel to, in order to generate a smoother field.
+
+        Returns
+        -------
+        cube : Iris.cube.Cube
+            Cube after applying a kernel, so that the resulting field is
+            smoothed.
+
+        """
+        try:
+            realiz_coord = cube.coord('realization')
+        except iris.exceptions.CoordinateNotFoundError:
+            pass
+        else:
+            if len(realiz_coord.points) > 1:
+                raise ValueError("Does not operate across realizations.")
+        if np.isnan(cube.data).any():
+            raise ValueError("Error: NaN detected in input cube data")
+
+        if self.lead_times is None:
+            radii_in_km = self.radii_in_km
+            ranges = self._get_grid_x_y_kernel_ranges(cube, radii_in_km)
+            cube = self._apply_kernel_for_smoothing(cube, ranges)
+        else:
+            required_lead_times = self._find_required_lead_times(cube)
+            # Interpolate to find the radius at each required lead time.
+            required_radii_in_km = (
+                np.interp(
+                    required_lead_times, self.lead_times, self.radii_in_km))
+            cubes = iris.cube.CubeList([])
+            # Find the number of grid cells required for creating the kernel,
+            # and then apply the kernel to smooth the field.
+            for cube_slice, radii_in_km in (
```

Comment:
This might make more sense as **radius_in_km** as each iteration is using a single radius.